planner/core/memtable_predicate_extractor.go

```diff
@@ -0,0 +1,200 @@
+	args := expr.GetArgs()
+	var col *expression.Column
+	var colIdx int
+	// c = 'rhs'
+	// 'lhs' = c
+	for i := 0; i < 2; i++ {
+		var isCol bool
+		col, isCol = args[i].(*expression.Column)
+		if isCol {
+			colIdx = i
+			break
+		}
+	}
+	if col == nil {
+		return "", nil
+	}
+
+	name, found := extractCols[col.UniqueID]
+	if !found {
+		return "", nil
+	}
+	// The `lhs/rhs` of EQ expression must be a constant
+	// SELECT * FROM t1 WHERE c='rhs'
+	// SELECT * FROM t1 WHERE 'lhs'=c
+	constant, ok := args[1-colIdx].(*expression.Constant)
+	if !ok || constant.DeferredExpr != nil || constant.ParamMarker != nil {
+		return "", nil
+	}
+	return name.ColName.L, []types.Datum{constant.Value}
+}
+
+func (helper extractHelper) intersection(lhs set.StringSet, datums []types.Datum, toLower bool) set.StringSet {
+	tmpNodeTypes := set.NewStringSet()
+	for _, datum := range datums {
+		var s string
+		if toLower {
+			s = strings.ToLower(datum.GetString())
+		} else {
+			s = datum.GetString()
+		}
+		tmpNodeTypes.Insert(s)
+	}
+	if len(lhs) > 0 {
+		return lhs.Intersection(tmpNodeTypes)
+	}
+	return tmpNodeTypes
+}
+
+// ClusterConfigTableExtractor is used to extract some predicates of `TIDB_CLUSTER_CONFIG`
+type ClusterConfigTableExtractor struct {
+	extractHelper
+
+	// SkipRequest means the where clause always false, we don't need to request any component
+	SkipRequest bool
+
+	// NodeTypes represents all components types we should send request to.
+	// e.g:
+	// 1. SELECT * FROM tidb_cluster_config WHERE type='tikv'
+	// 2. SELECT * FROM tidb_cluster_config WHERE type in ('tikv', 'tidb')
+	NodeTypes set.StringSet
+
+	// Addresses represents all components addresses we should send request to.
+	// e.g:
+	// 1. SELECT * FROM tidb_cluster_config WHERE address='192.168.1.7:2379'
+	// 2. SELECT * FROM tidb_cluster_config WHERE type in ('192.168.1.7:2379', '192.168.1.9:2379')
+	Addresses set.StringSet
+}
+
+// Extract implements the MemTablePredicateExtractor Extract interface
+func (e *ClusterConfigTableExtractor) Extract(schema *expression.Schema, names []*types.FieldName, predicates []expression.Expression) []expression.Expression {
+	remained := make([]expression.Expression, 0, len(predicates))
+	// All columns can be pushed down to the memory table `TIDB_CLUSTER_CONFIG`
+	const (
+		ColNameType    = "type"
+		ColNameAddress = "address"
+	)
+	extractCols := make(map[int64]*types.FieldName)
+	for i, name := range names {
+		if ln := name.ColName.L; ln == ColNameType || ln == ColNameAddress {
+			extractCols[schema.Columns[i].UniqueID] = name
+		}
+	}
+	// We don't support change the schema of memory table, so something must
+	// be wrong if the `type/address` columns cannot be found in the schema.
+	// The purpose of the following assert is used to make test failure because
+	// it will never happen if our release passing all test cases.
+	if len(extractCols) != 2 {
```

Comment:
I still don't understand the purpose of this check.

Reply:
I have reworked this comment, PTAL again.